### ISSUE TITLE
conform db access using ego.io

### ISSUE THREAD

**Issue by nesnoj** [openego/ego.io#16]
The original post was https://github.com/openego/data_processing/issues/13, further discussions can be found in https://github.com/openego/ego.io/issues/14#issuecomment-278313750:

> In either case we might aspire to bundle db access stuff in ego.io tools - currently db access is done using
> 1. oemof.db import statements (e.g. [here](https://github.com/openego/data_processing/blob/refactor/oedb-restructuring_v0.2/dataprocessing/python_scripts/peak_load_per_load_area.py#L19))
> 2. [data_proc io](https://github.com/openego/data_processing/blob/refactor/oedb-restructuring_v0.2/dataprocessing/tools/io.py) (e.g. [here](https://github.com/openego/data_processing/blob/refactor/oedb-restructuring_v0.2/dataprocessing/eGo_data_processing.py#L17))
> 3. [ego.powerflow tools](https://github.com/openego/ego.powerflow/blob/dev/egopowerflow/tools/tools.py) (e.g. [here](https://github.com/openego/ego.powerflow/blob/dev/egopowerflow/examples/hv_grid.py#L14))
> 
> We could move the `oedb_session()` function to ego.io and adjust all scripts in all repos to use it..?

This is on schedule for v0.3.

**Comment by nesnoj** [comment]
Done in #34

**Comment by mariusves** [comment]
I tried to use this connection provider with eTraGo so we could scrap the sessionmaker there.
Unfortunately the function `oemof_connection()` failed to retrieve my password from the keyring which made me enter it every time (maybe a windows issue?). Furthermore, the connection provider produces an error if a certain servername is not in the config file. I've made some changes to our sessionmaker [here](https://github.com/openego/eTraGo/tree/features/db-connection). It also includes a config-reader function similar to oemof.db, even though the password is not stored in the keyring but in the config file itself.

I'd propose the following "flow" for the connection provider:
1. Section (or servername) is present in config --> use config file
2. Section (or servername) is not present in config --> Ask if new connection details should be added to config or use them only once without saving
3. Config file not present --> use current implementation of create_oedb_config_file()
 
@nesnoj @gplssm @ulfmueller 
Any feedback on this? I would go ahead and implement the changes in a new branch to make it a bit clearer what my idea for this is.

**Comment by nesnoj** [comment]
> I tried to use this connection provider with eTraGo so we could scrap the sessionmaker there.
> Unfortunately the function oemof_connection() failed to retrieve my password from the keyring which made me enter it every time (maybe a windows issue?)

I do not know, it works well on Linux. Comments from windoze users, anyone?

> I'd propose the following "flow" for the connection provider:
> 1. Section (or servername) is present in config --> use config file
> 2. Section (or servername) is not present in config --> Ask if new connection details should be added to config or use them only once without saving
> 3. Config file not present --> use current implementation of create_oedb_config_file()

This sounds reasonable, please go ahead :+1: . But could you please implement it in ego.io (not eTraGo) and create a pull request? If you like, I could review the changes..
I propose to move the [part with the manual input](https://github.com/openego/ego.io/blob/b9d3153f02845110cc76a30f1a634e20ddeefef4/egoio/tools/db.py#L94-L122) to a separate function and use it during creation of a new file as well as during adding a new section.

Do you use `egoio.tools.db.connection()` in eTraGo or do you import directly from oemof?
(see [example in eDisGo](https://github.com/openego/eDisGo/blob/dev/edisgo/data/import_data.py#L1703)).

**Comment by mariusves** [comment]
With commit 8311a20efc274dd6e56316fa3d9dc40e0de89a17 I implemented my idea of a new config reader/writer.
![image](https://user-images.githubusercontent.com/20186954/33260460-3022d6f0-d360-11e7-8a72-f5412c62d871.png)

Please feel free to share your feedback!

**Comment by gplssm** [comment]
Thx for your effort! I would like to overcome the dependency of oemof.db with its archaic remnants just to initiate a DB connection ;-)

I haven't tried your code and I won't do it this week. But I'm highly interested to use it with ding0 and eDisGo!

If someone of you thinks about _how can I test all relevant use case?_, here's a list of things I would try if they work

1. Reading from a new section in the default config.ini (file already exists)
2. Reading from a new section in the default config.ini (file does not exist)
3. Reading from a new section in the non-default config.ini (use an arbitrary location, specified by the user). Write user input to that file
4. (optional) First check if connection details provided by input from prompt are valid and only save them if they are valid.

**Comment by nesnoj** [comment]
Thanks @mariusves for improving the db stuff, I'll test it within this week..

> I would like to overcome the dependency of oemof.db with its archaic remnants just to initiate a DB connection ;-)

I fully agree, time to get rid of it!

**Comment by gplssm** [comment]
Another note: currently when I use ego.io to create a database connection for the first time and enter the input for the config file, the password that I enter to be stored in the keyring in given plain text. Please adapt the password input to used getpass.

**Comment by nesnoj** [comment]
Yupp, same here. But getpass is already used, it says that it fall backs to plaintext for some reason..

**Comment by mariusves** [comment]
I've noticed that, too. Are you using Spyder/QTconsole, too? 
When searching for this issue I found the following issue [here](https://github.com/spyder-ide/spyder/issues/693) saying it was fixed in Spyder 2.3.9 & Qtconsole 4.2.1. Even though I'm using newer versions, this error still persists on my machine (win10 64bit).

**Comment by nesnoj** [comment]
W're both using pyCharm

**Comment by nesnoj** [comment]
@mariusves I've ran your version last week but got an error, so I decided not to dig deeper - this is what @gplssm mentioned in #41.
Now I reviewed your changes. In the end, the reason was a differently named key (db instead of database) - now it works.

But I fixed some small issues:
1. Rename db to database (more user-friendly from my pov)
2. Bugfix: Check if pass is None after retrieval of pass from keyring (in your version, [u used try..except](https://github.com/openego/ego.io/blob/e516cc55b30d65514bd804890564ea892664da08/egoio/tools/db.py#L212-L219) but if the password does not exist, keyring returns None instead of raising an exception!)
3. Removed `oemof.db` from setup.py, added `keyring`

I approved @gplssm's cases 1.-3. from above.

Merged into `dev`.
Thanks for your contribution!

**Comment by nesnoj** [comment]
If you're fine with this, please close @mariusves

**Comment by mariusves** [comment]
Thanks for the fixes and merging!

**Comment by ulfmueller** [comment]
Nice! Therefore eTraGo, eGo, eDisGo, dp (...) can (should) all now use this ego.io db connector?!
I suppose in each app are some (minor) adjustments necessary which all the individual app teams will manage, correct?!

**Comment by mariusves** [comment]
I can push the update to the new connector for eTraGo (already working with it locally).
